### ISSUE TITLE
Native support for warnings

### ISSUE THREAD

**Issue by Nutomic** [LemmyNet/lemmy#1424]
Mainly to help mods/admins to keep track of warnings. They should probably be shown in the user profile (only for mods/admins), and in the mod log (for everyone).

The warning should include a text, and probably a reference to the relevant post/comment.

**Comment by Nutomic** [comment]
For information, at the moment we are doing warnings simply by writing a comment to the offending user. This has the advantage that it is publicly visible directly in the thread, and I think it should stay that way. But at the same time, it would be nice if those warning comments were listed in the mod log and in the user profile (only visible for mods). Maybe it would be enough to add a comment field `is_warning`, which means the comment is a warning against the parent user.

**Comment by dessalines** [comment]
We now have a "distinguished" / "speak as mod" ability, so that might work for this, although there is currently no way to view those historically to see them as warnings.

**Comment by paxcoder** [comment]
please add this.
warnings would make it easier to moderate.

**Comment by iByteABit256** [comment]
>Maybe it would be enough to add a comment field is_warning, which means the comment is a warning against the parent user.

I see that the distinguished comment feature is very similar to this, should a new field be added for this or can the distinguished comment be modified to work as a warning?

I suppose this would also need queries like getting all warnings given to a specific target user, getting all communities based on the community or based on the moderator etc.

**Comment by Nutomic** [comment]
The way I see it this should be implemented by a new value `Warning` to the `ModlogKind` enum. Then the recipient will automatically receive a notification. It can also reference a specific post or comment via the `Modlog` table. 

This means warnings would be visible in the public modlog, but they could be filtered out if desired.

**Comment by dessalines** [comment]
I spose we basically just need a `warning: boolean` on create / edit comment. Then if its true (and if you're an admin / mod of that community), then create the modlog entry and the notification.

Side point but do we even really need distinguished? I've never really seen anyone use it, and mod / admin status is already shown in most UIs. 

The intention behind it was to "speak in an official capacity", but maybe its not really necessary.

**Comment by Nutomic** [comment]
You mean warnings should be made by posting a reply comment? That would be inconsistent with other mod actions. I was thinking a menu action similar to "remove comment" or "ban from community" with a dialog to enter the warning text. Then it will be sent to the user as notification automatically.

**Comment by iByteABit256** [comment]
The warning dialog action gives the flexibility that a mod can choose to give a warning without posting a public reply, and the distinguished flag can be used in combination with that if the moderator does want to give a public warning comment in addition

The UI can also have a checkbox under the warning reason that if checked, posts the same text as a distinguished comment

**Comment by dessalines** [comment]
Ah I suppose that makes more sense, as long as they could be federated.

**Comment by Nutomic** [comment]
Good point, Im not sure how this should work over federation. We should check how other platforms such as Mastodon federate warnings. Anyway it doesnt have to be part of the initial implementation.
